### ISSUE TITLE
OOME when running any app

### ISSUE THREAD

**Issue by OlegYch** [sbt/sbt#7403]
## steps

sbt 1.4.0 - 1.9.6
start sbt and `run` something in it
leave it running, doing nothing but waiting for input
get an OOME

## problem

`sbt.internal.util.Terminal.WriteableInputStream#readQueue` grows infinitely

## expectation

no OOME

## notes

a sample project triggering it (in a few hours) is here https://github.com/OlegYch/sbt-test-terminal-oome

**Comment by OlegYch** [comment]
![image](https://github.com/sbt/sbt/assets/268312/e01cfb16-e971-4030-bee0-bc0a3027cdbe)

**Comment by eed3si9n** [comment]
@OlegYch Thanks for the report! Not sure if this is relevant, but I see [WindowsInputStream](https://github.com/sbt/sbt/blob/1.10.x/internal/util-logging/src/main/scala/sbt/internal/util/WindowsInputStream.scala) as `in`, so I am assuming you're on some version of Windows?

**Comment by alexdupre** [comment]
The issue was reproduced on FreeBSD, Windows and Linux.

**Comment by eed3si9n** [comment]
@alexdupre Much appreciated!

**Comment by eed3si9n** [comment]
`readQueue` seems suspicious, and that appears in https://github.com/sbt/sbt/blob/1.10.x/internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala.

It sort of seems like `readQueue` is being used as a coordination mechanism between `read` method that writes `readQueue.put(())` and executor's thread that runs `impl()` in loop that calls `readQueue.take`? So if both `read` and the thread do no run at the same pace, `readQueue` could end up filling up with `()` value?

**Comment by OlegYch** [comment]
@eed3si9n exactly

**Comment by eed3si9n** [comment]
I guess we can put some upper bound on the writing side, and also clear out the queue once the reading is done.

**Comment by OlegYch** [comment]
this is too complex for me, but tinkering with bounds sounds like a stopgap :)

**Comment by eatkins** [comment]
I will explain the underlying design although I'm not sure exactly what is going on here. It is possible I have some details slightly wrong, but the spirit is correct:

sbt server is tricky because we need to support multiple clients potentially having an active ui session with the server. The server manages all of the client connections and creates a ui thread that reads from the client input stream and reacts. Remote clients are great because you can perform a cancellable blocking read on their input stream. The problem is that a blocking read from stdin is not cancellable. To make the embedded input stream of the process work the same as the remote connections, there is a global thread that reads from the process input stream. It tries to do this work lazily to avoid doing a blocking read that might not return. This is implemented by storing a buffer of bytes read from stdin. When calling read, we first check the buffer. If the buffer is empty, we write to the readQueue. This notifies the reading thread that it should perform a blocking read. But then it blocks on the buffer that is filled by the read thread. If the ui thread is not interrupted, the readQueue size should essentially never exceed 1.

To rephrase the above, we want to be able to do a blocking read from stdin when, and only when, the client asks for it. When the client asks for input and there is no buffered input available, then we trigger a blocking background read from stdin by writing to the readQueue (which activates a sleeping thread) and then block on the queue that is written to by the background thread. In theory, the ui thread should just be idling since sbt has nothing to do but wait for input so the readQueue should always be empty. The only way it can fill is if somehow the thread that is blocking waiting for the background thread is interrupted and https://github.com/sbt/sbt/blob/9f1d2fd4ed3ce95e7d0375f66eb6e17eac8aa954/internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala#L557 is reentered after the blocking read was already initiated. I don't understand why that is happening and it would be good to understand this before implementing a potential fix.

My suggestion is to figure out where https://github.com/sbt/sbt/blob/9f1d2fd4ed3ce95e7d0375f66eb6e17eac8aa954/internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala#L557
is being called. While idling, sbt should be blocked in read until there is input. Figuring out why that thread is being interrupted and restarted should hopefully lead to a proper fix. Trying to mess around with the bounds of the queue seems like a tempting but ultimately bad idea that could lead to bizarre flaky behavior. The readQueue is only meant to be a synchronization primitive. It should always be empty except for a brief moment when sbt initially enters an idle state and asks for a blocking read.

A possible fix would be to also add some additional state (one or two atomicbooleans) that prevent https://github.com/sbt/sbt/blob/9f1d2fd4ed3ce95e7d0375f66eb6e17eac8aa954/internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala#L557 from writing to the readQueue if the background thread is already performing a blocking read.

I am not going to write code or do any testing of my own, but I will be available to answer questions about the design.

**Comment by eed3si9n** [comment]
Thanks Ethan!

Since the repro code produced by Oleg is relatively short, I'll paste the entire content here:

```scala
object Main extends App {
  def run = while (true) {
    println(s"total - ${Runtime.getRuntime.totalMemory / 1024 / 1024} max - ${Runtime.getRuntime.maxMemory() / 1024 / 1024} free - ${Runtime.getRuntime.freeMemory() / 1024 / 1024}")
    Thread.sleep(30000)
  }

  new Thread(() => run){
    setDaemon(true)
    start()
  }
  Console.in.readLine()
}
```

Of the above, I'm guessing that the key ingredient is:

```scala
Console.in.readLine()
```

which is then blocked for X hours.

**Comment by eatkins** [comment]
I agree that is probably the key ingredient. It still doesn't explain to me why the Terminal/read method is not just blocking indefinitely and instead appears to be constantly interrupted and restarted somehow, filling the readQueue. This should be easy to trace though by adding `println(new Exception().getStackTrace().mkstring("\n"))` or something at the beginning of the Terminal/read method.

**Comment by eatkins** [comment]
I neglected to mention above that sbt does some crazy things to the global System.in/out in Terminal to make reading/writing from/to remote clients work without modifying application code (it's almost like global variables are a bad idea...). It is possible that Console.in is not well handled. I'm curious if System.in.readLine() reproduces the same behavior.

**Comment by eed3si9n** [comment]
Looks like `read` is called only once during the run?

```bash
[info] started sbt server
read
sbt:sbt-test-terminal-oome> impl
read
sbt:sbt-test-terminal-oome> impl
read
sbt:sbt-test-terminal-oome> impl
read
sbt:sbt-test-terminal-oome> runimpl

[info] running Main
total - 1963 max - 1963 free - 1770
read
total - 1963 max - 1963 free - 1767
total - 1963 max - 1963 free - 1758
total - 1963 max - 1963 free - 1754
total - 1963 max - 1963 free - 1753
total - 1963 max - 1963 free - 1753
total - 1963 max - 1963 free - 1753
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1752
total - 1963 max - 1963 free - 1751
```

**Comment by eatkins** [comment]
At least that is consistent with what I'd have expected. Is the readQueue growing in that test? I believe readQueue is only written to in one place so that should be easily observable in the same way.

**Comment by OlegYch** [comment]
@eatkins `System.in.read()` behaves in the same way

**Comment by eed3si9n** [comment]
I'm hoping that this would be fixed in sbt 2.0.0-RC8 (or any 2.x that uses client-side run).
